plugins/ccgx/fu-ccgx-dmc-device.c

```diff
@@ -0,0 +1,703 @@
+		return TRUE;
+
+	if (self->update_model == DMC_UPDATE_MODEL_DOWNLOAD_TIRGGER) {
+		DmcTriggerCode trigger_code = DMC_TRIGGER_CODE_UPDATE_NOW;
+
+		if (manual_replug)
+			trigger_code = DMC_TRIGGER_CODE_UPDATE_ON_DISCONNECT;
+
+		g_debug ("download tirgger:%u",trigger_code);
+
+		if (!fu_ccgx_dmc_device_send_download_trigger (self,
+							       trigger_code,
+							       error)) {
+			g_prefix_error (error, "download trigger error:");
+			return FALSE;
+		}
+	} else if (self->update_model == DMC_UPDATE_MODEL_PENDING_RESET) {
+
+		g_debug ("pending reset:%u",(guint32)manual_replug);
+
+		if (!fu_ccgx_dmc_device_send_sort_reset (self,
+							       manual_replug,
+							       error)) {
+			g_prefix_error (error, "soft reset error:");
+			return FALSE;
+		}
+	}  else {
+			g_set_error (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "invalid update model = %u",
+				     self->update_model);
+			return FALSE;
+	}
+
+	if (manual_replug)
+		return TRUE;
+
+	g_debug ("reset and wait replug");
+
+	remove_delay_ms = fu_device_get_install_duration(self) * 1000;
+	fu_device_set_remove_delay (FU_DEVICE (self), remove_delay_ms); /* over a minute to flash offline */
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_RESTART);
+	fu_device_add_flag (device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+	return TRUE;
+}
+
+static gboolean
+fu_ccgx_dmc_device_setup (FuDevice *device, GError **error)
+{
+	FuCcgxDmcDevice *self = FU_CCGX_DMC_DEVICE (device);
+	DmcDockStatus dock_status = {0};
+	DmcDockIdentity dock_id = {0};
+	guint32 version_raw = 0;
+	g_autofree gchar *version = NULL;
+	g_autofree gchar *instance_id = NULL;
+
+	/* get dock identity */
+	if (!fu_ccgx_dmc_device_get_dock_id (self, &dock_id, error))
+		return FALSE;
+
+	/* store dock identity */
+	if (!fu_memcpy_safe ((guint8 *)&self->dock_id, sizeof(DmcDockIdentity), 0x0, /* dst */
+		     (guint8*)&dock_id, sizeof(DmcDockIdentity), 0, sizeof(DmcDockIdentity), error)) /* src */
+		return FALSE;
+
+	/* get dock status */
+	if (!fu_ccgx_dmc_device_get_dock_status (self, &dock_status, error))
+		return FALSE;
+
+	/* set instance id */
+	instance_id = g_strdup_printf ("USB\\VID_%04X&PID_%04X&DMC",
+					fu_usb_device_get_vid (FU_USB_DEVICE (self)),
+					fu_usb_device_get_pid (FU_USB_DEVICE (self)));
+
+	fu_device_add_instance_id (FU_DEVICE (self), instance_id);
+
+	/* set composite version */
+	version_raw = dock_status.composite_version;
+	version = fu_ccgx_dmc_version_to_string (version_raw);
+	fu_device_set_version (FU_DEVICE (self), version);
+	fu_device_set_version_raw (FU_DEVICE (self), version_raw);
+
+	g_debug  ("[dmc dock] VID:%04X, PID:%04X, model:%d, vendor:%s, product:%s",
+		dock_id.vid, dock_id.pid, dock_id.model, dock_id.vendor_string, dock_id.product_string);
+
+	g_debug  ("[dmc status] status:0x%X, device count:%d, version:%s",
+		dock_status.device_status, dock_status.device_count, version);
```

Comment:
not helpful; if you want this make sure it's in the `_to_string()` handler.

Reply:
Removed on today commit (ccgx: fix code alignment, modify code for review).
Could you let me know how to see _to_sting() information?

Reply:
any time the daemon calls fu_device_to_string() -- so run the daemon with --verbose.